src/bindings/js/node/src/core_wrap.cpp

```diff
@@ -51,46 +56,216 @@ Napi::Value CoreWrap::read_model_async(const Napi::CallbackInfo& info) {
+        } else {
+            reportError(info.Env(), "Error while compiling model.");
+            return info.Env().Undefined();
+        }
+    } catch (std::exception& e) {
+        reportError(info.Env(), e.what());
+        return info.Env().Undefined();
+    }
+}
+
+struct TsfnContextModel {
+    TsfnContextModel(Napi::Env env) : deferred(Napi::Promise::Deferred::New(env)){};
+    std::thread nativeThread;
+
+    Napi::Promise::Deferred deferred;
+    Napi::ThreadSafeFunction tsfn;
+
+    std::shared_ptr<ov::Model> _model;
+    std::string _device;
+    ov::CompiledModel _compiled_model;
+    std::map<std::string, ov::Any> _config = {};
+};
+
+struct TsfnContextPath {
+    TsfnContextPath(Napi::Env env) : deferred(Napi::Promise::Deferred::New(env)){};
+    std::thread nativeThread;
+
+    Napi::Promise::Deferred deferred;
+    Napi::ThreadSafeFunction tsfn;
+
+    std::string _model;
+    std::string _device;
+    ov::CompiledModel _compiled_model;
+    std::map<std::string, ov::Any> _config = {};
+};
+
+void FinalizerCallbackModel(Napi::Env env, void* finalizeData, TsfnContextModel* context) {
+    context->nativeThread.join();
+    delete context;
+};
+
+void FinalizerCallbackPath(Napi::Env env, void* finalizeData, TsfnContextPath* context) {
+    context->nativeThread.join();
+    delete context;
+};
+
+void compileModelThreadModel(TsfnContextModel* context) {
+    ov::Core core;
+    context->_compiled_model = core.compile_model(context->_model, context->_device, context->_config);
+
+    auto callback = [](Napi::Env env, Napi::Function, TsfnContextModel* context) {
+        Napi::HandleScope scope(env);
+        auto obj = CompiledModelWrap::GetClassConstructor(env).New({});
+        auto cm = Napi::ObjectWrap<CompiledModelWrap>::Unwrap(obj);
+        cm->set_compiled_model(context->_compiled_model);
+
+        context->deferred.Resolve(obj);
+    };
+
+    context->tsfn.BlockingCall(context, callback);
+    context->tsfn.Release();
+}
+
+void compileModelThreadPath(TsfnContextPath* context) {
+    ov::Core core;
+    context->_compiled_model = core.compile_model(context->_model, context->_device, context->_config);
+
+    auto callback = [](Napi::Env env, Napi::Function, TsfnContextPath* context) {
+        Napi::HandleScope scope(env);
+        auto obj = CompiledModelWrap::GetClassConstructor(env).New({});
+        auto cm = Napi::ObjectWrap<CompiledModelWrap>::Unwrap(obj);
+        cm->set_compiled_model(context->_compiled_model);
+
+        context->deferred.Resolve(obj);
+    };
+
+    context->tsfn.BlockingCall(context, callback);
+    context->tsfn.Release();
+}
+
+Napi::Value CoreWrap::compile_model_async(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+    if (info[0].IsObject() && info[1].IsString()) {
+        auto context_data = new TsfnContextModel(env);
+        auto m = Napi::ObjectWrap<ModelWrap>::Unwrap(info[0].ToObject());
+        context_data->_model = m->get_model()->clone();
```

Comment:
It could be expensive operation

Reply:
Here we can either create a copy of a model or use some kind of mechanism to synchronize access, because model will be shared between threads

Reply:
Please, leave a FIXME with related comment